distributed/process.py

```diff
@@ -119,39 +125,26 @@ def _start():
             state.pid = process.pid
             logger.debug("[%s] created process with pid %r" % (r, state.pid))
 
-        def _process_one_message(timeout):
-            try:
-                msg = q.get(timeout=timeout)
-            except Empty:
-                pass
-            else:
-                logger.debug("[%s] got message %r" % (r, msg))
-                op = msg['op']
-                if op == 'start':
-                    _call_and_set_future(loop, msg['future'], _start)
-                elif op == 'terminate':
-                    _call_and_set_future(loop, msg['future'], process.terminate)
-                elif op == 'stop':
-                    raise SystemExit
-                else:
-                    assert 0, msg
-
-        def _maybe_notify_exit(exitcode):
-            self = selfref()  # only keep self alive when required
-            try:
-                if self is not None:
-                    self._loop.add_callback(self._on_exit, exitcode)
-            finally:
-                self = None  # lose reference
-
         while True:
             # Periodic poll as there's no simple way to poll a threading Queue
             # and a mp Process at the same time.
-            try:
-                _process_one_message(timeout=0.02)
-            except SystemExit:
-                return
-            # Did process end?
+            msg = q.get()
+            logger.debug("[%s] got message %r" % (r, msg))
+            op = msg['op']
+            if op == 'start':
+                _call_and_set_future(loop, msg['future'], _start)
+            elif op == 'terminate':
+                _call_and_set_future(loop, msg['future'], process.terminate)
+            elif op == 'stop':
+                break
+            else:
+                assert 0, msg
+
+    @classmethod
+    def _watch_process(cls, selfref, process, state, q):
+        r = repr(selfref())
+        while True:
```

Comment:
We now use process.join *after* we poll for when the process starts.  We could probably do this more cleanly, but I'm going to punt for now.